AssertionError when integrating cos(a*x+b) where x is a BaseScalar / Coordinate Variable

Attempting to use integrate(cos(a*x+b),(x,xl,xu)) where x is a BaseScalar fails with an assertion error in heurisch.py. This works fine when x is a symbol (i.e., x=symbols('x')). The documentation (https://docs.sympy.org/latest/modules/vector/fields.html) does imply that the use of the BaseScalar / Coordinate variable in this context should be valid (...BaseScalar instances can be used just like any other SymPy Symbol...).

Code: 
from sympy import *
from sympy.vector import CoordSys3D, dot, cross
init_printing()
N = CoordSys3D('N')
x = N.x
a,b,xu,xl = symbols('a,b,xu,xl')
integrate(cos(a*x+b),(x,xl,xu))

Output: 
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-3-cf475f300897> in <module>
      7 x = N.x
      8 a,b,xu,xl = symbols('a,b,xu,xl')
----> 9 integrate(cos(a*x+b),(x,xl,xu))

~/.local/share/virtualenvs/modes-1_lY0sjy/lib/python3.5/site-packages/sympy/integrals/integrals.py in integrate(*args, **kwargs)
   1463     if isinstance(integral, Integral):
   1464         return integral.doit(deep=False, meijerg=meijerg, conds=conds,
-> 1465                              risch=risch, manual=manual)
   1466     else:
   1467         return integral

~/.local/share/virtualenvs/modes-1_lY0sjy/lib/python3.5/site-packages/sympy/integrals/integrals.py in doit(self, **hints)
    536                 else:
    537                     antideriv = self._eval_integral(
--> 538                         function, xab[0], **eval_kwargs)
    539                     if antideriv is None and meijerg is True:
    540                         ret = try_meijerg(function, xab)

~/.local/share/virtualenvs/modes-1_lY0sjy/lib/python3.5/site-packages/sympy/integrals/integrals.py in _eval_integral(self, f, x, meijerg, risch, manual, conds)
    987                 try:
    988                     if conds == 'piecewise':
--> 989                         h = heurisch_wrapper(g, x, hints=[])
    990                     else:
    991                         h = heurisch(g, x, hints=[])

~/.local/share/virtualenvs/modes-1_lY0sjy/lib/python3.5/site-packages/sympy/integrals/heurisch.py in heurisch_wrapper(f, x, rewrite, hints, mappings, retries, degree_offset, unnecessary_permutations)
    130 
    131     res = heurisch(f, x, rewrite, hints, mappings, retries, degree_offset,
--> 132                    unnecessary_permutations)
    133     if not isinstance(res, Basic):
    134         return res

~/.local/share/virtualenvs/modes-1_lY0sjy/lib/python3.5/site-packages/sympy/integrals/heurisch.py in heurisch(f, x, rewrite, hints, mappings, retries, degree_offset, unnecessary_permutations)
    451     if mappings is None:                                                #
    452         # optimizing the number of permutations of mapping              #
--> 453         assert mapping[-1][0] == x  # if not, find it and correct this comment
    454         unnecessary_permutations = [mapping.pop(-1)]
    455         mappings = permutations(mapping)

AssertionError: 


Environment: 
jeremy@Newton:~/modes$ pipenv run pip freeze
backcall==0.1.0
bleach==3.0.2
decorator==4.3.0
defusedxml==0.5.0
entrypoints==0.2.3
ipykernel==5.1.0
ipython==7.1.1
ipython-genutils==0.2.0
ipywidgets==7.4.2
jedi==0.13.1
Jinja2==2.10
jsonschema==2.6.0
jupyter==1.0.0
jupyter-client==5.2.3
jupyter-console==6.0.0
jupyter-core==4.4.0
MarkupSafe==1.1.0
mistune==0.8.4
mpmath==1.0.0
nbconvert==5.4.0
nbformat==4.4.0
notebook==5.7.2
numpy==1.15.4
pandocfilters==1.4.2
parso==0.3.1
pexpect==4.6.0
pickleshare==0.7.5
prometheus-client==0.4.2
prompt-toolkit==2.0.7
ptyprocess==0.6.0
Pygments==2.2.0
python-dateutil==2.7.5
pyzmq==17.1.2
qtconsole==4.4.3
scipy==1.1.0
Send2Trash==1.5.0
six==1.11.0
sympy==1.3
terminado==0.8.1
testpath==0.4.2
tornado==5.1.1
traitlets==4.3.2
wcwidth==0.1.7
webencodings==0.5.1
widgetsnbextension==3.4.2
jeremy@Newton:~/modes$ pipenv run python3 --version
Python 3.5.2

--- a/sympy/integrals/heurisch.py
+++ b/sympy/integrals/heurisch.py
@@ -57,7 +57,7 @@ def components(f, x):
     result = set()
 
     if x in f.free_symbols:
-        if f.is_Symbol:
+        if f.is_symbol and f.is_commutative:
             result.add(f)
         elif f.is_Function or f.is_Derivative:
             for g in f.args:
@@ -450,6 +450,7 @@ def heurisch(f, x, rewrite=False, hints=None, mappings=None, retries=3,
     rev_mapping = {v: k for k, v in mapping}                            #
     if mappings is None:                                                #
         # optimizing the number of permutations of mapping              #
+        assert mapping[-1][0] == x # if not, find it and correct this comment
         unnecessary_permutations = [mapping.pop(-1)]
         mappings = permutations(mapping)
     else:

--- a/sympy/integrals/heurisch.py
+++ b/sympy/integrals/heurisch.py
@@ -450,7 +450,6 @@ def heurisch(f, x, rewrite=False, hints=None, mappings=None, retries=3,
     rev_mapping = {v: k for k, v in mapping}                            #
     if mappings is None:                                                #
         # optimizing the number of permutations of mapping              #
-        assert mapping[-1][0] == x  # if not, find it and correct this comment
         unnecessary_permutations = [mapping.pop(-1)]
         mappings = permutations(mapping)
     else:

--- a/sympy/integrals/tests/test_integrals.py
+++ b/sympy/integrals/tests/test_integrals.py
@@ -16,7 +16,7 @@ from sympy.utilities.pytest import XFAIL, raises, slow, skip, ON_TRAVIS
 from sympy.utilities.randtest import verify_numerically
 from sympy.integrals.integrals import Integral
 
-x, y, a, t, x_1, x_2, z, s = symbols('x y a t x_1 x_2 z s')
+x, y, a, t, x_1, x_2, z, s, b= symbols('x y a t x_1 x_2 z s b')
 n = Symbol('n', integer=True)
 f = Function('f')
 
@@ -1465,6 +1465,11 @@ def test_issue_15640_log_substitutions():
     f = sqrt(log(x))/x**2
     F = -sqrt(pi)*erfc(sqrt(log(x)))/2 - sqrt(log(x))/x
     assert integrate(f, x) == F and F.diff(x) == f
+
+def test_issue_15509():
+    assert integrate(cos(a*x + b), (x, x_1, x_2)) == Piecewise(
+        (-sin(a*x_1 + b)/a + sin(a*x_2 + b)/a, (a > -oo) & (a < oo) & Ne(a, 0)), \
+            (-x_1*cos(b) + x_2*cos(b), True))
 
 
 def test_issue_4311():

--- a/sympy/integrals/tests/test_integrals.py
+++ b/sympy/integrals/tests/test_integrals.py
@@ -1467,7 +1467,10 @@ def test_issue_15640_log_substitutions():
     assert integrate(f, x) == F and F.diff(x) == f
 
 def test_issue_15509():
-    assert integrate(cos(a*x + b), (x, x_1, x_2)) == Piecewise(
+    from sympy.vector import CoordSys3D    
+    N = CoordSys3D('N')
+    x = N.x
+    assert integrate(cos(a*x + b), (x, x_1, x_2), heurisch = True) == Piecewise(
         (-sin(a*x_1 + b)/a + sin(a*x_2 + b)/a, (a > -oo) & (a < oo) & Ne(a, 0)), \
             (-x_1*cos(b) + x_2*cos(b), True))
 

--- a/sympy/integrals/tests/test_integrals.py
+++ b/sympy/integrals/tests/test_integrals.py
@@ -1467,10 +1467,10 @@ def test_issue_15640_log_substitutions():
     assert integrate(f, x) == F and F.diff(x) == f
 
 def test_issue_15509():
-    from sympy.vector import CoordSys3D    
+    from sympy.vector import CoordSys3D
     N = CoordSys3D('N')
     x = N.x
-    assert integrate(cos(a*x + b), (x, x_1, x_2), heurisch = True) == Piecewise(
+    assert integrate(cos(a*x + b), (x, x_1, x_2), heurisch=True) == Piecewise(
         (-sin(a*x_1 + b)/a + sin(a*x_2 + b)/a, (a > -oo) & (a < oo) & Ne(a, 0)), \
             (-x_1*cos(b) + x_2*cos(b), True))
 

--- a/sympy/vector/scalar.py
+++ b/sympy/vector/scalar.py
@@ -43,6 +43,7 @@ class BaseScalar(AtomicExpr):
         return obj
 
     is_commutative = True
+    is_symbol = True
 
     @property
     def free_symbols(self):

